src/plugins/Workers/TransactionSyncStreamWorker/methods/startHistoricalSync.js

```diff
@@ -0,0 +1,61 @@
+const GrpcErrorCodes = require('@dashevo/grpc-common/lib/server/error/GrpcErrorCodes');
+
+const logger = require('../../../../logger');
+
+const GRPC_RETRY_ERRORS = [
+  GrpcErrorCodes.DEADLINE_EXCEEDED,
+  GrpcErrorCodes.UNAVAILABLE,
+  GrpcErrorCodes.INTERNAL,
+  GrpcErrorCodes.CANCELLED,
+  GrpcErrorCodes.UNKNOWN,
+];
+
+/**
+ *
+ * @param {string} network
+ * @return {Promise<void>}
+ */
+module.exports = async function startHistoricalSync(network) {
+  const lastSyncedBlockHash = this.getLastSyncedBlockHash();
+  const bestBlockHeight = await this.getBestBlockHeightFromTransport();
+  const lastSyncedBlockHeight = await this.getLastSyncedBlockHeight();
+  const count = bestBlockHeight - lastSyncedBlockHeight || 1;
+  const start = +new Date();
+
+  try {
+    const options = { count, network };
+    // If there's no blocks synced, start from height 0, otherwise from the last block hash.
+    if (lastSyncedBlockHash == null) {
+      options.fromBlockHeight = lastSyncedBlockHeight;
+    } else {
+      options.fromBlockHash = lastSyncedBlockHash;
+    }
+
+    logger.debug(`TransactionSyncStreamWorker - HistoricalSync - Started from ${options.fromBlockHash || options.fromBlockHeight}, count: ${count}`);
+    const gapLimitIsReached = await this.syncUpToTheGapLimit(options);
+    if (gapLimitIsReached) {
+      await startHistoricalSync.call(this, network);
+    }
+  } catch (e) {
+    if (GRPC_RETRY_ERRORS.includes(e.code)) {
+      if (this.stream === null) {
```

Comment:
Cancel throws some specific error right? Shouldn't we check it?

Reply:
No, it doesn't, it just an error with `.code` property. It's the same in the dapi-client: https://github.com/dashevo/js-dapi-client/blob/88dab147c5a07e7a20ca04c5713de6fc5ff27d62/lib/transport/GrpcTransport.js#L71